File: lib/extensions/track.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:metadata_god/metadata_god.dart';
import 'package:path/path.dart';
import 'package:spotify/spotify.dart';
import 'package:spotube/provider/spotify/spotify.dart';
import 'package:spotube/services/audio_player/audio_player.dart';
import 'package:spotube/services/logger/logger.dart';

extension TrackExtensions on Track {
  Track fromFile(
    File file, {
    Metadata? metadata,
    String? art,
  }) {
    album = Album()
      ..name = metadata?.album ?? "Unknown"
      ..images = [if (art != null) Image()..url = art]
      ..genres = [if (metadata?.genre != null) metadata!.genre!]
      ..artists = [
        Artist()
          ..name = metadata?.albumArtist ?? "Unknown"
          ..id = metadata?.albumArtist ?? "Unknown"
          ..type = "artist",
      ]
      ..id = metadata?.album
      ..releaseDate = metadata?.year?.toString();
    artists = [
      Artist()
        ..name = metadata?.artist ?? "Unknown"
        ..id = metadata?.artist ?? "Unknown"
    ];

    id = metadata?.title ?? basenameWithoutExtension(file.path);
    name = metadata?.title ?? basenameWithoutExtension(file.path);
    type = "track";
    uri = file.path;
    durationMs = (metadata?.durationMs?.toInt() ?? 0);

    return this;
  }

  Metadata toMetadata({
    required int fileLength,
    Uint8List? imageBytes,
  }) {
    return Metadata(
      title: name,
      artist: artists?.map((a) => a.name).join(", "),
      album: album?.name,
      albumArtist: artists?.map((a) => a.name).join(", "),
      year: album?.releaseDate != null
          ? int.tryParse(album!.releaseDate!.split("-").first) ?? 1969
          : 1969,
      trackNumber: trackNumber,
      discNumber: discNumber,
      durationMs: durationMs?.toDouble() ?? 0.0,
      fileSize: BigInt.from(fileLength),
      trackTotal: album?.tracks?.length ?? 0,
      picture: imageBytes != null
          ? Picture(
              data: imageBytes,
              // Spotify images are always JPEGs
              mimeType: 'image/jpeg',
            )
          : null,
    );
  }
}

extension IterableTrackSimpleExtensions on Iterable<TrackSimple> {
  Future<List<Track>> asTracks(AlbumSimple album, ref) async {
    try {
      final spotify = ref.read(spotifyProvider);
      final tracks = await spotify.invoke(
        (api) => api.tracks.list(map((trackSimple) => trackSimple.id!).toList()));
      return tracks.toList();
    } catch (e, stack) {
      // Ignore errors and create the track locally
      AppLogger.reportError(e, stack);

      List<Track> tracks = [];
      for (final trackSimple in this) {
        Track track = Track();
        track.album = album;
        track.name = trackSimple.name;
        track.artists = trackSimple.artists;
        track.availableMarkets = trackSimple.availableMarkets;
        track.discNumber = trackSimple.discNumber;
        track.durationMs = trackSimple.durationMs;
        track.explicit = trackSimple.explicit;
        track.externalUrls = trackSimple.externalUrls;
        track.href = trackSimple.href;
        track.id = trackSimple.id;
        track.isPlayable = trackSimple.isPlayable;
        track.linkedFrom = trackSimple.linkedFrom;
        track.previewUrl = trackSimple.previewUrl;
        track.trackNumber = trackSimple.trackNumber;
        track.type = trackSimple.type;
        track.uri = trackSimple.uri;
        tracks.add(track);
      }
      return tracks;
    }
  }
}

extension TracksToMediaExtension on Iterable<Track> {
  List<SpotubeMedia> asMediaList() {
    return map((track) => SpotubeMedia(track)).toList();
  }
}

File: android/app/src/main/kotlin/oss/krtirtho/spotube/glance/models/Track.kt
package oss.krtirtho.spotube.glance.models

import com.google.gson.annotations.SerializedName
import kotlinx.serialization.Serializable
import kotlin.time.Duration.Companion.milliseconds

@Serializable
data class Track(
    val album: AlbumSimple?, val artists: List<Artist>?,

    @SerializedName("available_markets") val availableMarkets: List<Market>?,

    @SerializedName("disc_number") val discNumber: Int?,

    @SerializedName("duration_ms") val durationMs: Int,

    val explicit: Boolean?, val href: String?, val id: String?,

    @SerializedName("is_playable") val isPlayable: Boolean?,

    val name: String?,

    @SerializedName("popularity") val popularity: Int?,

    @SerializedName("preview_url") val previewUrl: String?,

    @SerializedName("track_number") val trackNumber: Int?,

    val type: String?, val uri: String?
) {
    val duration: kotlin.time.Duration
        get() = durationMs.toLong().milliseconds
}

enum class Market {
    AD, AE, AF, AG, AI, AL, AM, AO, AQ, AR, AS, AT, AU, AW, AX, AZ, BA, BB, BD, BE, BF, BG, BH, BI, BJ, BL, BM, BN, BO, BQ, BR, BS, BT, BV, BW, BY, BZ, CA, CC, CD, CF, CG, CH, CI, CK, CL, CM, CN, CO, CR, CU, CV, CW, CX, CY, CZ, DE, DJ, DK, DM, DO, DZ, EC, EE, EG, EH, ER, ES, ET, FI, FJ, FK, FM, FO, FR, GA, GB, GD, GE, GF, GG, GH, GI, GL, GM, GN, GP, GQ, GR, GS, GT, GU, GW, GY, HK, HM, HN, HR, HT, HU, ID, IE, IL, IM, IN, IO, IQ, IR, IS, IT, JE, JM, JO, JP, KE, KG, KH, KI, KM, KN, KP, KR, KW, KY, KZ, LA, LB, LC, LI, LK, LR, LS, LT, LU, LV, LY, MA, MC, MD, ME, MF, MG, MH, MK, ML, MM, MN, MO, MP, MQ, MR, MS, MT, MU, MV, MW, MX, MY, MZ, NA, NC, NE, NF, NG, NI, NL, NO, NP, NR, NU, NZ, OM, PA, PE, PF, PG, PH, PK, PL, PM, PN, PR, PS, PT, PW, PY, QA, RE, RO, RS, RU, RW, SA, SB, SC, SD, SE, SG, SH, SI, SJ, SK, SL, SM, SN, SO, SR, SS, ST, SV, SX, SY, SZ, TC, TD, TF, TG, TH, TJ, TK, TL, TM, TN, TO, TR, TT, TV, TW, TZ, UA, UG, UM, US, UY, UZ, VA, VC, VE, VG, VI, VN, VU, WF, WS, XK, YE, YT, ZA, ZM, ZW,
}
